### PR TITLE
Route rate limits (sec #18)

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

Implements the rate-limiting part of [sec #18](https://github.com/flowforge/security/issues/18)

### Commits: 
* add rateLimits.js and required plugin
* setup routes for inclusion/exclusion of limiting
* login and account create pages handle 429 errors
* add tests and required fastify-routes plugin
* add default config to ff yaml file
* update docs for rate-limiting
* sync package lock
* add upgrade section with link to new docs

**_This PR is probably best reviewed by commit as they were made in logical order_**

### Implementation detail
1. Adds `/forge/routes/rateLimits.js`
    * This has one function: `getLimits` which returns the rate limit settings. It is directly exported for unit testing.
2. `/config/index.js`
    * Adds `rate_limits` to `config` with default values by calling to `rateLimits.getLimits(...)`
3. `/forge/forge.js`
    * Registers `@fastify/rate-limit` plugin - **but only if `config.rate_limits.enabled` is `true`**
4. Any routes that should be rate limited have `config: { rate_limits: app.config.rate_limits }` added to their route options.
5. Any routes that should not be rate limited have `config: { rate_limits: false }` added to their route options.

### Endpoint Rate Limiting

Rate limiting will be disabled by default.  

To enable rate limiting, set the config setting `rate_limits.enabled` to `true`.

#### When enabled, the following config settings are available:
* `rate_limits.global` - (boolean, default: `true`) - enable/disable rate limiting for all routes
* `rate_limits.timeWindow` - (number, default: `"1 minute"`) - time window in milliseconds for `rate_limits.max` requests
* `rate_limits.max` - (number, default: `1000`) - max number of requests per `rate_limits.timeWindow` for a given route
* `rate_limits.maxAnonymous` - (number, default: _not set_) - max number of requests per `rate_limits.timeWindow` for a given route for anonymous users
* _Additional settings are possible - see [@fastify/rate-limit#options](https://github.com/fastify/fastify-rate-limit#options) for more details._

##### The following routes will be rate limited by default:
* POST /account/login
* POST /account/register
* POST /account/forgot_password
* POST /account/reset_password/:token

##### The following routes will never be rate limited:
* POST /api/comms/auth/client
* POST /api/comms/auth/acl
* POST /account/logout
* POST /account/verify/:token
* POST /account/verify
* GET /account/authorize
* POST /account/token
* PUT /api/v1/devices/:deviceId/editor
* GET /api/v1/devices/:deviceId/editor
* GET /api/v1/devices/:deviceId/editor/token
* GET /api/v1/devices/:deviceId/editor/comms/:access_token
* GET /api/v1/devices/:deviceId/editor/proxy/*
* POST /api/v1/devices/:deviceId/editor/proxy/*
* PUT /api/v1/devices/:deviceId/editor/proxy/*
* HEAD /api/v1/devices/:deviceId/editor/proxy/*
* DELETE /api/v1/devices/:deviceId/editor/proxy/*
* OPTIONS /api/v1/devices/:deviceId/editor/proxy/*

##### When config setting `rate_limits.global` is `true`.
* Everything not explicitly excluded will be rate limited.

##### When config setting `rate_limits.global` is `true`.
* Nothing else will be rate limited

### package.json changes
* Adds the following to package.json
    * dependencies
        * @fastify/rate-limit
    * devDependencies
        * @fastify/routes

#### Important notes

* In order to get access the fastify routes, the plugin @fastify/routes added to the dev dependencies.
    * It is only loaded when the setting `options.config.test?.fastifyRoutes` is `true`.
    * NOTE: access to app.routes is very useful - may want to consider loading the routes plugin by default.

### Adds tests in `/test/unit/forge/routes/api/rateLimits/rateLimits_spec.js`

NOTE: There are 3 sections of tests under " Rate Limiting Routes" for `global:true` and `global:false`:
1. routes that are explicitly enabled (and should always have limiting regardless of `global` setting)
2. routes that are explicitly disabled (and should never be limited regardless of `global` setting)
3. a random selection of endpoints that should be enabled by `global:true`, otherwise disabled.

#### Test results (local execution)
```
Endpoint Rate Limiting
    getLimits unit tests
      ✔ should be disabled by default
      ✔ should get correct default values
      ✔ max should be a function when maxAnonymous is set
      ✔ keyGenerator should return sid if logged in, ip if not
      ✔ keyGenerator should return x-real-ip first, then x-forwarded-for, then ip
    Rate Limiting plugin
      ✔ Plugin is not loaded when config settings are not specified (implicitly disabled) (1038ms)
      ✔ Plugin is not loaded when config settings explicitly disables rate limits (693ms)
      ✔ Plugin is loaded when config settings are enabled (724ms)
    Rate Limiting Routes
      All Routes use limits except explicitly excluded (global:true)
        Routes that should always be limited
          ✔ Route POST /account/login should be rate limited
          ✔ Route POST /account/register should be rate limited
          ✔ Route POST /account/forgot_password should be rate limited
          ✔ Route POST /account/reset_password/:token should be rate limited
        Routes that should never be limited
          ✔ Route POST /api/comms/auth/client should never be rate limited
          ✔ Route POST /api/comms/auth/acl should never be rate limited
          ✔ Route POST /account/logout should never be rate limited
          ✔ Route POST /account/verify/:token should never be rate limited
          ✔ Route POST /account/verify should never be rate limited
          ✔ Route GET /account/authorize should never be rate limited
          ✔ Route POST /account/token should never be rate limited
          ✔ Route PUT /api/v1/devices/:deviceId/editor should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor/token should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor/comms/:access_token should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route POST /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route PUT /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route HEAD /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route DELETE /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route OPTIONS /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
        Routes that should be limited because the global is true
          ✔ Route GET / should be rate limited
          ✔ Route GET /app/main.js should be rate limited
          ✔ Route GET /admin/overview should be rate limited
          ✔ Route GET /api/v1/teams/:teamId/members should be rate limited
          ✔ Route GET /api/v1/applications/:applicationId should be rate limited
          ✔ Route GET /api/v1/projects/:instanceId/logs should be rate limited
      Only specific Routes use limits (global:false)
        Routes that should always be limited
          ✔ Route POST /account/login should be rate limited
          ✔ Route POST /account/register should be rate limited
          ✔ Route POST /account/forgot_password should be rate limited
          ✔ Route POST /account/reset_password/:token should be rate limited
        Routes that should never be limited
          ✔ Route POST /api/comms/auth/client should never be rate limited
          ✔ Route POST /api/comms/auth/acl should never be rate limited
          ✔ Route POST /account/logout should never be rate limited
          ✔ Route POST /account/verify/:token should never be rate limited
          ✔ Route POST /account/verify should never be rate limited
          ✔ Route GET /account/authorize should never be rate limited
          ✔ Route POST /account/token should never be rate limited
          ✔ Route PUT /api/v1/devices/:deviceId/editor should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor/token should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor/comms/:access_token should never be rate limited
          ✔ Route GET /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route POST /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route PUT /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route HEAD /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route DELETE /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
          ✔ Route OPTIONS /api/v1/devices/:deviceId/editor/proxy/* should never be rate limited
        Routes that should not be limited because global is false
          ✔ Route GET / should not be rate limited
          ✔ Route GET /app/main.js should not be rate limited
          ✔ Route GET /admin/overview should not be rate limited
          ✔ Route GET /api/v1/teams/:teamId/members should not be rate limited
          ✔ Route GET /api/v1/applications/:applicationId should not be rate limited
          ✔ Route GET /api/v1/projects/:instanceId/logs should not be rate limited
```

## Related Issue(s)

https://github.com/flowforge/security/issues/18

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [x] Upgrade instructions
    - [x] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: docs/install/configuration.md [hunk 1]
```diff
@@ -138,6 +138,26 @@ Option        | Description
 `telemetry.frontend.posthog.capture_pageview` | FlowForge is designed as to provide custom posthog `$pageview` events that provide more detail on navigation than the default, and suit a single page application better. As such, we recommend setting this to false in order to prevent duplicate `pageleave`/`pageview` events firing. Default: `true`
 
 
+## Rate Limiting configuration
+
+By default, rate limiting is disabled and the platform will not rate limit any requests.
+
+To enable rate limiting, you can set the `rate_limits.enabled` option to `true`. This will enable rate limiting for all routes unless explicitly disabled (see note below).
+By default, all routes will be limited requests to 1000 per 1 minute window for logged in users, and 60 per 1 minute window for anonymous users.
```
Comment (Steve-Mcl) [inline docs/install/configuration.md:146]:
```suggestion
By default, all routes will be limited to 1000 per 1 minute window.
```

Comment (Steve-Mcl) [inline docs/install/configuration.md:146]:
```suggestion
By default, rate limiting is disabled and the platform will not rate limit any requests.

To enable rate limiting, you can set the `rate_limits.enabled` option to `true`. 

When enabled, all routes will be limited to 1000 requests per 1 minute window. These defaults can be adjusted by setting values in the configuration options listed below.
```

File: frontend/src/pages/account/Create.vue [hunk 2]
```diff
@@ -170,9 +184,15 @@ export default {
                         this.errors.email = err.response.data.error
                     } else if (/email/.test(err.response.data.error)) {
                         this.errors.email = 'Email unavailable'
-                    }
-                    if (err.response.data.error === 'user registration not enabled') {
+                    } else if (err.response.data.statusCode === 429) {
+                        this.errors.general = 'Too many login attempts. Try again later.'
```
Comment (Steve-Mcl) [inline frontend/src/pages/account/Create.vue:188]:
```suggestion
                        this.errors.general = 'Too many attempts. Try again later.'
```

File: docs/install/configuration.md [hunk 3]
```diff
@@ -138,6 +138,27 @@ Option        | Description
...
 
+## Rate Limiting configuration
+
+By default, rate limiting is disabled and the platform will not rate limit any requests.
+
+To enable rate limiting, you can set the `rate_limits.enabled` option to `true`. 
+
+When enabled, all routes will be limited to 1000 requests per 1 minute window. These defaults can be adjusted by setting values in the configuration options listed below.
+
+NOTE: Some routes are prohibited from being rate limited, such as the logout and acl routes.
```
Comment (hardillb) [inline docs/install/configuration.md:149]: While I understand why this is documented I wonder if it just makes it a target for attack

Comment (Steve-Mcl) [reply docs/install/configuration.md:149]: Good question Ben.  Open to suggestions.

Comment (Steve-Mcl) [inline docs/install/configuration.md:149]:
```suggestion
```


@hardillb - remove this section?

Something else?

Comment (Steve-Mcl) [reply docs/install/configuration.md:149]: will re-do.

[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply docs/install/configuration.md:149]: @hardillb 🤞

**Review by hardillb — APPROVED** [review]

**Pull request merged**